android/src/main/java/com/polidea/flutter_ble_lib/SafeMainThreadResolver.java

```diff
@@ -0,0 +1,44 @@
+package com.polidea.flutter_ble_lib;
+
+import android.os.Handler;
+import android.os.Looper;
+
+import com.polidea.multiplatformbleadapter.OnErrorCallback;
+import com.polidea.multiplatformbleadapter.OnSuccessCallback;
+import com.polidea.multiplatformbleadapter.errors.BleError;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class SafeMainThreadResolver<T> {
+
+    private OnErrorCallback onErrorCallback = null;
+    private OnSuccessCallback<T> onSuccessCallback = null;
+    private AtomicBoolean called = new AtomicBoolean(false);
+
+    public SafeMainThreadResolver(OnSuccessCallback<T> onSuccessCallback, OnErrorCallback onErrorCallback) {
+        this.onErrorCallback = onErrorCallback;
+        this.onSuccessCallback = onSuccessCallback;
+    }
+
+    public void onSuccess(final T data) {
+        if (onSuccessCallback != null && called.compareAndSet(false, true)) {
+            new Handler(Looper.getMainLooper()).post(new Runnable() {
```

Comment:
This kind of construction may lead to a memory leak. I tried to trace all usages and it seems to be ok.  DO we have anywhere access to activity?

Reply:
No, the library uses only application context.